src/webgpu/api/operation/rendering/stencil.spec.ts

```diff
@@ -0,0 +1,230 @@
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    const depthStencilFormat: GPUTextureFormat = 'depth24plus-stencil8';
+    const depthTexture = this.device.createTexture({
+      size: { width: 1, height: 1, depthOrArrayLayers: 1 },
+      format: depthStencilFormat,
+      sampleCount: 1,
+      mipLevelCount: 1,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_DST,
+    });
+
+    const depthStencilAttachment: GPURenderPassDepthStencilAttachment = {
+      view: depthTexture.createView(),
+      depthLoadOp: 'load',
+      depthStoreOp: 'store',
+      stencilLoadOp: 'load',
+      stencilStoreOp: 'store',
+    };
+
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: renderTarget.createView(),
+          storeOp: 'store',
+          loadOp: 'load',
+        },
+      ],
+      depthStencilAttachment,
+    });
+
+    // Draw the base triangle with stencil reference 1.
+    // This clears the stencil buffer to 1.
+    {
+      const testPipeline = this.createRenderPipelineForTest(baseState);
+      pass.setPipeline(testPipeline);
+      pass.setStencilReference(1);
+      pass.setBindGroup(
+        0,
+        this.createBindGroupForTest(
+          testPipeline.getBindGroupLayout(0),
+          new Float32Array(kBaseColor)
+        )
+      );
+      pass.draw(1);
+    }
+
+    // Draw a triangle with the given stencil reference and the comparison function.
+    // The color will be stencilColor if the stencil test passes, and kBaseColor if not.
+    {
+      const testPipeline = this.createRenderPipelineForTest(state);
+      pass.setPipeline(testPipeline);
+      pass.setStencilReference(stencilRefValue);
+      pass.setBindGroup(
+        0,
+        this.createBindGroupForTest(
+          testPipeline.getBindGroupLayout(0),
+          new Float32Array(stencilColor)
+        )
+      );
+      pass.draw(1);
+    }
+
+    pass.end();
+    this.device.queue.submit([encoder.finish()]);
+
+    this.expectSinglePixelIn2DTexture(
+      renderTarget,
+      renderTargetFormat,
+      { x: 0, y: 0 },
+      { exp: new Uint8Array(stencilColor) }
+    );
+  }
+
+  createRenderPipelineForTest(depthStencil: GPUDepthStencilState): GPURenderPipeline {
+    return this.device.createRenderPipeline({
+      layout: 'auto',
+      vertex: {
+        module: this.device.createShaderModule({
+          code: `
+            @vertex
+            fn main(@builtin(vertex_index) VertexIndex : u32) -> @builtin(position) vec4<f32> {
+                return vec4<f32>(0.0, 0.0, 0.0, 1.0);
+            }
+            `,
+        }),
+        entryPoint: 'main',
+      },
+      fragment: {
+        targets: [{ format: 'rgba8unorm' }],
+        module: this.device.createShaderModule({
+          code: `
+            struct Params {
+              color : vec3<f32>
```

Comment:
The bind group is still getting initialized with 4 components (kBaseColor or kStencilColor) which means the alpha channel is dropped (and would end up in `depth` if you still had it here). 
How about just changing this to vec4 to avoid that?

Reply:
ah right, it's looks like changing it to vec4 is more flexible for the the next tests.